### ISSUE TITLE
Detected an invalid packet at item 0

### ISSUE THREAD

**Issue by mathursiddarth** [anastas/gr-cdma#12]
Hello anastas,
When running the "cdma_txrx", I keep getting the "INFO: Detected an invalid packet at item 0". What does this imply? Is it like you mentioned before that it does not find a valid packet header?
Thank you.

**Comment by anastas** [comment]
This is perfectly reasonable:
the receiver is trying to synchronize with the incoming samples.
He thinks there is a packet starting at 0 but after demodulating the header he finds that the header is erroneous and so he tells you so...

**Comment by anastas** [comment]
In real life you would have a higher layer that would 
1) either discard this packet or
2) send a NACK to the Tx to retransmit it

I have only designed the PHY layer, so I don't need to worry about what to do with this message...

**Comment by mathursiddarth** [comment]
Ok. Yeah I get what you are saying here. Thank you.

**Comment by mathursiddarth** [comment]
Got another thing I want to know. I am running "cdma_tx" and "cdma_rx" on separate USRPs and I don't see anything on the receiver. I increased the gain as well. Which other factor could be affecting the reception here?

Also, doing the same with "cdma_tx1" and "cdma_rx1", I do get something on the receiver side but majority of them are detected as an invalid packet something like this:

"EC82AD9E3E46C99EA4BD5BDE333CDF9CEE13F902EE82745D2CCCF8D4575148B18942470B0A45EAE7AAF104DA41A52A672D

INFO: Detected an invalid packet at item 70416
INFO: Detected an invalid packet at item 71208

EC82AD9E3E46C99EA4BD5BDE333CDF9CEE13F902EE82745D2CCCF8D4575148B18942470B0A45EAE7AAF104DA41A52A672D

INFO: Detected an invalid packet at item 72576
INFO: Detected an invalid packet at item 76248
INFO: Detected an invalid packet at item 76392
INFO: Detected an invalid packet at item 76824.."

I have put an "Any Block Sink" at the decoded_payload_bytes to see what I am receiving and the "EC82AD9E..." is a result of that.

Thank you for the help.

**Comment by anastas** [comment]
There are a lot of things that can go wrong when you are running through actual USRP's if you do not customize your gr-cdma PHY layer appropriately.
The most sensitive part is frequency/timing acquisition.
You have to have an idea of how your chosen carrier frequency, chip rate, daughterboard clock stability, etc affect the frequency drift. 
This does not mean that gr-cdma is not designed to cope with that; it just means that you need to parameterize it appropriately (eg, by increasing the number of filters in the frequency filterbank).

My suggestion to you it to start slowly by first running experiments with the two USRP's fully synchronized by an external clock so that you eliminate that piece of uncertainty.
Then you have to appropriately parameterize your cdma for the environment (including specific USRP/daughterboards) you are using.

This process is common to any PHY layer; not just gr-cdma...

**Comment by mathursiddarth** [comment]
Hi Anastas,
I tried a few things here, changing the parameters. I tried some different transmissions on the two USRPs and they seem to be working alright. Attaching a few screenshots of what I see at the output of few of the blocks in gr-cdma, comparing cdma_txrx1 and cdma_rx(Over the air reception). Can you help me out here with what could possible be happening? I do see reception but it is quite garbled up. Thank you.
![chopper_input_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230806/d66e2b78-d974-11e5-82fd-89cbfda25816.png)
![chopper_input_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230807/d66e4220-d974-11e5-8dc3-7ec1b1661d57.png)
![corrected_out_data_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230804/d66dfb30-d974-11e5-9f26-a3cc6f70dcef.png)
![corrected_out_data_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230808/d66e902c-d974-11e5-8c16-5d52753c346b.png)
![header_corrected_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230805/d66e120a-d974-11e5-95bb-4b519ce7d82e.png)
![header_corrected_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230809/d66eaf80-d974-11e5-9b14-f8a18155f4ba.png)
![payload_corrected_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230810/d673b69c-d974-11e5-9763-1f73b5e15bec.png)
![payload_corrected_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230811/d675f39e-d974-11e5-8b74-9b2c6450163a.png)

**Comment by mathursiddarth** [comment]
Here is this RX_samples plotted at the receiver.

**Comment by anastas** [comment]
To start:

please download the LATEST code and work ALWAYS with the cdma_tx and cdma_rx files.
There is no point complicating the design by using the adaptive-modulation version.
Once you have a working non-adaptive system you can move on to the adaptive one.

Then:
Make sure that the parameter n_filt is large enough so that the parameter f_max
(shown on your RX window when you run it)
is at least as large as the expected frequency drift from the clocks of you daughterboard.

Then:
Observe the system in operation:
If it frequently goes from tracking to acquisition to tracking etc then the system is not very stable,
but if it stays in tracking mode for some time then things should be OK.

Then:
take a look at the estimated Es/N0 (it is shown on the RX window when running)
If the Es/N0 data estimated is ~10dB it means that BPSK or QPSK should be running with very good reliability (BER=10^-5 for Eb/N0=9.6 dB)

**Comment by mathursiddarth** [comment]
After git pull now, it says "cp is not defined". Could it be due to older version of gnuradio? Because I did the same on another computer with the latest gnu and it worked fine.

**Comment by anastas** [comment]
Hey look,
I am not going to answer such questions on your part.

What does it mean: " it says "cp is not defined""
who "says" that?
have you done a clean install?
have you tried to resolve the issue yourself?
have you provided me or others with information about the apparent bug?
or do you think we are here to serve you?

if you need help you have to do better than that...

**Comment by mathursiddarth** [comment]
Hey. What I meant by "It says" was that most of the blocks are running the error and cannot recognize the source of cp.  Yes, I did do a clean install and did try to resolve the issue myself. The only reason I asked you was to know if you had an idea of where the issue lies and if the version of GRC could affect. It's absolutely upto you if you want to answer that or not. In no way did I mean to hurt your sentiments like that. I apologize if it came out that way.

**Comment by anastas** [comment]
open python.
Do 
> from cdma import cdma_parameters as cp

if this gives you a problem, then your installation has a problem.
Have you changed the "prefix" inside the cdma_parameters.py file?

**Comment by mathursiddarth** [comment]
I get this error with the python command:

"  File "/usr/local/lib/python2.7/dist-packages/cdma/__init__.py", line 34, in <module>
    import cdma_parameters
  File "/usr/local/lib/python2.7/dist-packages/cdma/cdma_parameters.py", line 115, in <module>
    payload_mod = [digital.constellation_qpsk(),digital.constellation_8psk(),digital.constellation_16qam()]
AttributeError: 'module' object has no attribute 'constellation_16qam'".

Removing the "constellation_16qam" parameters from the cdma_parameters.py did solve the problem temporarily though.
I did change the prefix in that file.

**Comment by anastas** [comment]
Indeed, this means you are running an older gnuradio that does not have the 16QAM modulation code.
This is not important though, since as I mentioned earlier you should not run the adaptive modulation version, but only the standard cdma_tx/rx code.
You did the right thing to remove it from the code for now...

**Comment by mathursiddarth** [comment]
Yes. Thank you for confirming that.
Also, I did run the cdma_tx/rx code and

1. It stayed in acquisition mode for the whole time. Never switched to Tracking. Certainly means it is not stable at all.
2. Estimated Es/N0 is a negative value every time I run it.

**Comment by anastas** [comment]
both observations are inconsistent with your previous plots:
eg, your 1st plot shows "mode=tracking" and  "training Es/N0 ~10.4dB"

In any case your current system design does not seem stable.

**Comment by mathursiddarth** [comment]
Those plots were for cdma_rx1. Getting different results for cdma_rx.

**Comment by anastas** [comment]
This is impossible!
The two systems are the same;
Cdma_tx1/rx1 has the option of adapting the mod but you are not using it...

Are you sure you are running both with the same parameters?

**Comment by mathursiddarth** [comment]
Yes. The parameters are exactly the same. I made no changes in the parameters. And yes, I am sure I am running both the cdma_tx and cdma_rx.
I don't understand how cdma_tx1/rx1 seem to be working but with cdma_rx I see negative Es/N0.
I tried with the fifo files and it is working correctly with "traning Es/N0 ~ 37 dB". Although I see constant switching between Acquision/Tracking when with fifos.

**Comment by mathursiddarth** [comment]
Scratch the Acquisition/Tracking switching. That was due to the Repeat in file. So with fifos, it is working perfectly fine.

**Comment by anastas** [comment]
This requires further investigation as it is highly suspicious.
(maybe we have made some change in cdma_tx/rx that broke things).

In that case, you can switch to the cdma_tx1/rx1 which works for you...

**Comment by anastas** [comment]
If you want to help us figuring out that, could you do some kind of binary search and figure out at which point the cdma_tx/rx is broken, by going backwards and testing previous commits...

that would be very useful to pinpoint the error (if there is any).

**Comment by mathursiddarth** [comment]
Ok sure. I will give that a shot and let you know if I find anything. Thank you so much.

**Comment by mathursiddarth** [comment]
Another thing with the cdma_tx1/rx1 is that there are a lot of packets that are not detected properly. Like I have posted before in the thread, there are innumerable number of packets which go detected invalidly.

**Comment by mathursiddarth** [comment]
Hey Anastas,
I tried with a different set of USRPs and the cdma_tx/rx seem to be working now. So, we know for sure the code is not broken. The "Training Es/N0 ~ 10dB". 
Although, I do see frequent switchings between Acquisition/Tracking and a lot of packets detected as invalid like in the case of cdma_tx1/rx1.

**Comment by anastas** [comment]
So, starting with the assumption that the two systems behave "similarly",
the next step is to concentrate on tx/rx and improve the system design.

Can you verify that your CPU is NOT overtwhelmed during acquisition?
One way to do that is to switch to "manual" and then switch to "ack".
Then make sure that your system is not at 100%.

Also make sure you do not see dropped packets from the USRP.

Finally make sure that the USRP does not give you any kind of WARNING when you start your code
regarding the sample rate (this depends on the symbol rate you have chosen etc).

If all these do not improve performance, you need to design a system with a SMALLER frame size.

**Comment by mathursiddarth** [comment]
CPU is overwhelmed at the receiver. The utilization is ~200% during acquisition.

I don't think there are any dropped packets from the USRP. Can you tell me how to verify that?

And no, there are no warnings.

**Comment by anastas** [comment]
So if you have a multi-core CPU then 200% is not bad.
During acquisition a LARGE number (n_filt=11) of long filters is running 
which is usually the bottleneck. However due to the thread-per-block architecture of GNURADIO
this task is automatically distributed to different threads/cores.

If you want to improve the performance you can try several things:

1) Increase the n_filt to a bigger value (maybe 21 or more-- as much as your multi-core CPU can handle) so that you can cope with the frequency drift coming from USRPs
If you want to eliminate this problem so that you can check if this indeed is the issue, you can synchronize the two USRPs using an external clock. If your system performance improves considerably then indeed the issue is frequency drift

2) Also this problem can be aleviated (not eliminated) by having smaller frames.

**Comment by mathursiddarth** [comment]
So, you are saying the bottleneck caused by the large number of filters can lead the invalid detection of packets at the receiver? I changed the number to 150 in the GRC block and nothing changed. I still see frequent switching between Ack/Tracking and the CPU utilization is the same. Where do I exactly change the n_filt value? I am not sure if I am doing it right.

I will try and do the synchronization using an external clock and see if that works out.

Also, can you help me understand how can I exactly change the design for smaller frames? Just need to know how to go about it.

Thank you.

**Comment by anastas** [comment]
what exactly did you change to 150?

all these parameters are set in cdma_parameters.py

and are documented in the documentation of these files...

**Comment by mathursiddarth** [comment]
I have 8 cores. I was doing it wrong and after changing the n_filt to 21, the utilization got up to 280%. Should I try and utilize as many cores as I can? I am not sure if I see a bottleneck. How do I know if there is one? I apologize about all the questions. I am just new to this. Thanks.

**Comment by anastas** [comment]
The bigger n_filt the bigger the parameter f_max (that appears on the RX window).
With n_filt=21 probably your f_max = 2.5KHz (is this correct?)
If you make it 41 it will go to 5 kHz and so on (with an almost proportional corresponding increase in the CPU load). I suggest you set it to a value where you are roughly at 80 to 90% of total CPU load (which is 800% in your 8-core machine).

But even that may not be enough...
You have to email to the discuss-gnuradio list (or better yet to the usrp-users list) and ask what is your
expected frequency discrepancy between TX and RX for the specific sample rate and center frequency you are using at your USRP/daughterboard.
If this lies within the +-f_max range then you are OK.

Another way to test this is to eliminate this discrepance by using a common clock and see what happens (as I mentioned above)

A third way is to use a simple test GRC system that transmits a simple sinusoid at some frequency
(say 20KHz) above your carrier frequency and observe it at the RX with an FFT scope.
You will see that it moves around. Now you can estimate roughly how far away it moves (and how fast). This will give you an indication of the system you are coping with.

**Comment by mathursiddarth** [comment]
I have set n_filt to 150 and the load is around 680% (85% of the load) with the f_max = 18.625 KHz.

I have also e-mailed the usrp-users list to get the expected frequency drift for my USRP/daughterboard.

The third way you mentioned, I did transmit a simple cosine wave at 20KHz on 2.2G. I do not see the center frequency move at all. If it is, it certainly is less than 18KHz which is our f_max. Am I looking at it correctly?

![cosinefft2](https://cloud.githubusercontent.com/assets/17165617/13337268/787108bc-dbe8-11e5-82c2-ad86f64926c4.png)
![fft cosine](https://cloud.githubusercontent.com/assets/17165617/13337269/787256c2-dbe8-11e5-833e-cbca862e53a4.png)

**Comment by anastas** [comment]
what is the difference between the two plots?

**Comment by mathursiddarth** [comment]
FFT size is 1024. Both plots are the same. Just took the screenshots at different times.
![cosinescopesink](https://cloud.githubusercontent.com/assets/17165617/13338399/1ed91824-dbef-11e5-9504-ad7865c371b8.png)
![cosinetx](https://cloud.githubusercontent.com/assets/17165617/13338398/1ed8cd4c-dbef-11e5-94de-031e406ff4fc.png)

**Comment by anastas** [comment]
could you please attach a figure of your GRC system (not the signals, but the actual block diagram...)

**Comment by mathursiddarth** [comment]
I will do that tonigh. I was working on a remote machine and don't have access to it now. It was just the sine wave to the USRP and the FFT Sink connected to the USRP Sink at the receiver .

**Comment by mathursiddarth** [comment]
![transmitter](https://cloud.githubusercontent.com/assets/17165617/13344267/85388b26-dc22-11e5-8c97-010c42575f32.png)

![receiver](https://cloud.githubusercontent.com/assets/17165617/13344266/852fb758-dc22-11e5-87b6-13a17c74e1de.png)

**Comment by anastas** [comment]
ok, so let me ask something:
when you run tx/rx, how many errors do you see (roughly) over a 10 second period?

**Comment by mathursiddarth** [comment]
About 24 of these. With n_filt = 150.

_INFO: Detected an invalid packet at item 0
INFO: Detected an invalid packet at item 36
INFO: Detected an invalid packet at item 72
INFO: Detected an invalid packet at item 108
INFO: Detected an invalid packet at item 180
INFO: Detected an invalid packet at item 360
INFO: Detected an invalid packet at item 576
INFO: Detected an invalid packet at item 720
INFO: Detected an invalid packet at item 1908
INFO: Detected an invalid packet at item 2484
INFO: Detected an invalid packet at item 12636
INFO: Detected an invalid packet at item 12672
INFO: Detected an invalid packet at item 17424
INFO: Detected an invalid packet at item 17460
INFO: Detected an invalid packet at item 17532
INFO: Detected an invalid packet at item 17568
INFO: Detected an invalid packet at item 17604
INFO: Detected an invalid packet at item 17712
INFO: Detected an invalid packet at item 17784
INFO: Detected an invalid packet at item 17820
INFO: Detected an invalid packet at item 17928
INFO: Detected an invalid packet at item 17964
INFO: Detected an invalid packet at item 18108
INFO: Detected an invalid packet at item 30348_

**Comment by anastas** [comment]
So all this discussion this last week is about 24 erroneous packets per 10 seconds?
Do you realize what is he corresponding packet error rate ?

**Comment by mathursiddarth** [comment]
This got better once we increased the n_filt value to 150. And it is different every time I run it as well, it lags due to the high load on the CPU. Sometimes it is as high as 50-75.

**Comment by anastas** [comment]
Do you realize what is he corresponding packet error rate ?

**Comment by mathursiddarth** [comment]
10^-3 to 10^-5?

**Comment by anastas** [comment]
what is the packet (frame) rate of your system?

**Comment by mathursiddarth** [comment]
I calculated the packet error rate according to how many packets were actually dropped in the first 30k packets which is about 8x10^-4.
I am not sure about the packet(frame) rate of the system. Not able to figure it out.

**Comment by anastas** [comment]
can you verify the following in your setup:

The measured SNr starts at sime high value (say 12 or 14) and then within a period of 4-5 seconds it drops to below 10 at which point an acquisition is trigered and then the process starts again.

is this what you observe?

can you describe what you see (or post a short video of it somewhere?)

**Comment by mathursiddarth** [comment]
Yes. That is what is happening. It goes from positive to negative every time acquisition is triggered.
I will post a video in a while too.

**Comment by mathursiddarth** [comment]
https://drive.google.com/file/d/0B3D0ZMGZtx0LRDY4QTdTa2dNNGc/view?usp=docslist_api

**Comment by mathursiddarth** [comment]
https://drive.google.com/file/d/0B3D0ZMGZtx0LejdMMzZnb0RKWmc/view?usp=sharing

**Comment by mathursiddarth** [comment]
So, do you think that there is something wrong with the design here?

**Comment by anastas** [comment]
this points to the instability of clocks as i pointed out earlier.

You can try feeding your Tx/Rx the same reference clock and see if this solves the problem.
(please let us know if this solves the problem!)

Clearly we can improve the design to do tracking so that this phenomenon is aleviated.

**Comment by anastas** [comment]
I run the following experiment that confirms that the clock drift is the culprit:

I now have both USRPs connected to my PC 
(in the network interface p6p2 IP= 192.168.30.1):

USRP #1  is 192.168.30.3
and
USRP #2 is 192.168.30.5

Only #1 is connected with an Ethernet cable to PC while the other one (#2)
is connected to #1 via the MIMO cable.

Then I run 
cdma_tx with TX being USRP #1 (setting the clok/tim to "default")
and
cdma_rx with RX being USRP #2 (setting the clock/time to "MIMO cable")

and I get a VERY STABLE system: it keeps running in tracking mode for 
more than 5 mins now (with SNR at 16dB)...


This means that really I have to work on the tracking part to make this system 
stable for the case of non synced clocks!!!

But at least now we know what is going on...

**Comment by lbb123** [comment]
hello anastas,
   I just want to know where is  spread spectrum code and modulation  in your cdma system?Because I want to recongnise the type of modulation automaticly.Tahk you~

looking for your repaly

**Comment by lbb123** [comment]
In other words, I see your python code in cdma_tx_hier.py.I found 'cp.payload_mod[0].points()'  in th file,but I think it just constellation mapping not modulation.

**Comment by lbb123** [comment]
And I mean the sine wave modulation.

**Comment by anastas** [comment]
The constellation points {+-1} for the chips are pulse-shaped with the transmitted filter (after up-sampling) and this constitutes the baseband modulated signal.
In gnuradio we do not have to generate sine waves (ie, the RF signal) in software.
This is done by the USRP device!

**Comment by lbb123** [comment]
Thank you very much,anastas. 
Another question, the spread spectrum code is not been considered in your system,is right?

**Comment by anastas** [comment]
I do not understand the question!
The ss code(s) are fully customizable by the user.

**Comment by Keskebeu** [comment]
I will be joining gr-cdma as well I am waiting for a clock reference to be delivered from china so i can synch my two radios. Hopefully soon. I like the project.

**Comment by shrin18** [comment]
> Hi Anastas,
> I tried a few things here, changing the parameters. I tried some different transmissions on the two USRPs and they seem to be working alright. Attaching a few screenshots of what I see at the output of few of the blocks in gr-cdma, comparing cdma_txrx1 and cdma_rx(Over the air reception). Can you help me out here with what could possible be happening? I do see reception but it is quite garbled up. Thank you.
> ![chopper_input_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230806/d66e2b78-d974-11e5-82fd-89cbfda25816.png)
> ![chopper_input_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230807/d66e4220-d974-11e5-8dc3-7ec1b1661d57.png)
> ![corrected_out_data_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230804/d66dfb30-d974-11e5-9f26-a3cc6f70dcef.png)
> ![corrected_out_data_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230808/d66e902c-d974-11e5-8c16-5d52753c346b.png)
> ![header_corrected_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230805/d66e120a-d974-11e5-95bb-4b519ce7d82e.png)
> ![header_corrected_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230809/d66eaf80-d974-11e5-9b14-f8a18155f4ba.png)
> ![payload_corrected_cdmarx1](https://cloud.githubusercontent.com/assets/17165617/13230810/d673b69c-d974-11e5-9763-1f73b5e15bec.png)
> ![payload_corrected_cdmatxrx1](https://cloud.githubusercontent.com/assets/17165617/13230811/d675f39e-d974-11e5-8b74-9b2c6450163a.png)

Hey @mathursiddarth  can you provide a Screenshot of the block diagram of the implementation with the USRP. W are trying to replicate the project using Pluto SDR, however we have the source input block as complex float 32 and so changed throttle to complex. However we are not able to connect cdma_tx_hier to throttle since it accepts Integer input.

**Comment by anastas** [comment]
unless you have a fully functional cdma_txrx AND you understand what it does, there is not point in moving to either USRP implementaiton, or the more complicated system with different modulations (ie, the application files that end in "1")...
